jupyterhub_files/spawner.py

```diff
@@ -359,20 +365,38 @@ def remote_notebook_start(self, instance):
         for key in env:
             lenv = lenv + key + "=" + env[key] + " "
         # End setup environment
-        self.log.debug("function remote_server_start %s" % self.user.name)
+        self.log.debug("function remote_notebook_start %s" % self.user.name)
         worker_ip_address_string = instance.private_ip_address
         start_notebook_cmd = self.cmd + self.get_args()
         start_notebook_cmd = " ".join(start_notebook_cmd)
-        self.log.info("Starting user %s jupyterhub" % self.user.name)
+        self.log.info("Starting user %s private ip %s jupyterhub" % (self.user.name, worker_ip_address_string))
         with settings(user = self.user.name, key_filename = FABRIC_DEFAULTS["key_filename"],  host_string=worker_ip_address_string):
-             yield sudo("%s %s --user=%s --notebook-dir=/home/%s/ --allow-root > /tmp/jupyter.log 2>&1 &" % (lenv, start_notebook_cmd,self.user.name,self.user.name),  pty=False)
-        self.log.debug("just started the notebook for user %s, waiting." % self.user.name)
-        try:
-            self.user.settings[self.user.name] = instance.public_ip_address
-        except:
-            self.user.settings[self.user.name] = ""
+            yield sudo("%s %s --user=%s --notebook-dir=/home/%s/ --allow-root > /tmp/jupyter.log 2>&1 &" % (lenv, start_notebook_cmd,self.user.name,self.user.name),  pty=False)
+            self.log.debug("just started the notebook for user %s, private ip %s, waiting." % (self.user.name, worker_ip_address_string))
+            try:
+                self.user.settings[self.user.name] = instance.public_ip_address
+            except:
+                self.user.settings[self.user.name] = ""
+            notebook_running = yield self.is_notebook_running(worker_ip_address_string, attempts=30)
+            self.log.debug("value for variable notebook_running in function remote_notebook_start for user %s private ip %s is %s" % (self.user.name, worker_ip_address_string, notebook_running))
+            num_remote_notebook_start_retries = 0
+            while not notebook_running and num_remote_notebook_start_retries < REMOTE_NOTEBOOK_START_RETRY_MAX:
+                yield sudo("%s %s --user=%s --notebook-dir=/home/%s/ --allow-root > /tmp/jupyter.log 2>&1 &" % (lenv, start_notebook_cmd,self.user.name,self.user.name),  pty=False)
+                self.log.debug("just retried to start the notebook for user %s, private ip %s. Attempt %s. Waiting..." % (self.user.name, worker_ip_address, num_remote_notebook_start_retries + 1))
+                try:
+                    self.user.settings[self.user.name] = instance.public_ip_address
+                except:
+                    self.user.settings[self.user.name] = ""
+                notebook_running = yield self.is_notebook_running(worker_ip_address_string, attempts=30)
```

Comment:
If I'm understanding the process correctly, a complete and total failure to start the notebook would result in 5 * 30 = 150 attempts. Prior to this change, we tried to start the notebook server once and then retried up to 30 times. 

The assumption is that we can SSH into the server, but the remote notebook server might fail to start and so we need to retry a few times, with a delay in between. To that end, I'm wondering if we could achieve a similar effect by reducing the number of attempts when checking if the notebook server is running and instead add a short delay after those attempts (e.g. `yield gen.sleep(10)`). This would provide a little more time for the server to start without incurring additional network traffic, which could help at peak times. 

Having said that, consider this for a future enhancement if the extra network traffic turns out to be a problem.

Reply:
Makes total sense. I'll just go ahead and implement that right away.